File: tier2_intermediate/divide_and_conquer/solutions/p427_construct_quad_tree.c
/*
 * P427: Construct Quad Tree [PREMIUM] (Medium)
 * https://leetcode.com/problems/construct-quad-tree/
 * Topics: Array, Divide and Conquer, Tree, Matrix
 *
 * Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.
 * Return the root of the Quad-Tree representing grid.
 *
 * Example 1:
 *     Input: grid = [[0,1],[1,0]]
 *     Output: [[0,1],[1,0],[1,1],[1,1],[1,0]]
 *
 * Constraints:
 *     - n == grid.length == grid[i].length
 *     - n == 2x where 0 <= x <= 6
 *
 * Hint: Recursively divide grid into 4 quadrants, make leaf if all same value.
 */


#include "io.h"
#include <stdlib.h>
#include <string.h>

typedef struct QNode {
    int is_leaf;
    int val;
    struct QNode *tl, *tr, *bl, *br;
} QNode;

static QNode *make_leaf(int v) {
    QNode *n = calloc(1, sizeof(QNode));
    n->is_leaf = 1; n->val = v;
    return n;
}

static QNode *make_internal(QNode *tl, QNode *tr, QNode *bl, QNode *br) {
    QNode *n = calloc(1, sizeof(QNode));
    n->is_leaf = 0; n->val = 1;
    n->tl = tl; n->tr = tr; n->bl = bl; n->br = br;
    return n;
}

static int all_same(int *grid, int n, int row, int col, int size) {
    int v = grid[row * n + col];
    for (int i = row; i < row + size; i++)
        for (int j = col; j < col + size; j++)
            if (grid[i * n + j] != v) return 0;
    return 1;
}

static QNode *build(int *grid, int n, int row, int col, int size) {
    if (all_same(grid, n, row, col, size))
        return make_leaf(grid[row * n + col]);
    int half = size / 2;
    QNode *tl = build(grid, n, row, col, half);
    QNode *tr = build(grid, n, row, col + half, half);
    QNode *bl = build(grid, n, row + half, col, half);
    QNode *br = build(grid, n, row + half, col + half, half);
    return make_internal(tl, tr, bl, br);
}

static void free_tree(QNode *node) {
    if (!node) return;
    free_tree(node->tl); free_tree(node->tr);
    free_tree(node->bl); free_tree(node->br);
    free(node);
}

static void print_level_order(QNode *root) {
    if (!root) return;
    typedef struct { QNode **data; int len, cap, head; } Queue;
    Queue q;
    q.cap = 64; q.data = malloc(q.cap * sizeof(QNode*)); q.len = 0; q.head = 0;
    q.data[q.len++] = root;
    while (q.head < q.len) {
        QNode *node = q.data[q.head++];
        if (!node) { printf("null\n"); continue; }
        printf("%d %d\n", node->is_leaf, node->val);
        if (!node->is_leaf) {
            if (q.len + 4 > q.cap) { q.cap *= 2; q.data = realloc(q.data, q.cap * sizeof(QNode*)); }
            q.data[q.len++] = node->tl;
            q.data[q.len++] = node->tr;
            q.data[q.len++] = node->bl;
            q.data[q.len++] = node->br;
        }
    }
    free(q.data);
}

int main(void)
{
    int cn;
    int *cdata = read_ints(&cn);
    int cols = cdata[0];
    free(cdata);

    int *grid = malloc(cols * cols * sizeof(int));
    for (int i = 0; i < cols; i++) {
        int rn;
        int *row = read_ints(&rn);
        memcpy(grid + i * cols, row, rn * sizeof(int));
        free(row);
    }

    QNode *root = build(grid, cols, 0, 0, cols);
    print_level_order(root);
    free_tree(root);
    free(grid);
    return 0;
}

File: tier2_intermediate/divide_and_conquer/solutions/p427_construct_quad_tree.rs
/*
 * P427: Construct Quad Tree [PREMIUM] (Medium)
 * https://leetcode.com/problems/construct-quad-tree/
 * Topics: Array, Divide and Conquer, Tree, Matrix
 *
 * Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.
 * Return the root of the Quad-Tree representing grid.
 * A Quad-Tree is a tree data structure in which each internal node has exactly four children. Besides, each node has two attributes:
 * We can construct a Quad-Tree from a two-dimensional area using the following steps:
 * If you want to know more about the Quad-Tree, you can refer to the wiki.
 * Quad-Tree format:
 * You don't need to read this section for solving the problem. This is only if you want to understand the output format here. The output represents the serialized format of a Quad-Tree using level order traversal, where null signifies a path terminator where no node exists below.
 * It is very similar to the serialization of the binary tree. The only difference is that the node is represented as a list [isLeaf, val].
 * If the value of isLeaf or val is True we represent it as 1 in the list [isLeaf, val] and if the value of isLeaf or val is False we represent it as 0.
 *     class Node {
 *     public boolean val;
 *     public boolean isLeaf;
 *     public Node topLeft;
 *     public Node topRight;
 *     public Node bottomLeft;
 *     public Node bottomRight;
 *     }
 *
 * Example 1:
 *     Input: grid = [[0,1],[1,0]]
 *     Output: [[0,1],[1,0],[1,1],[1,1],[1,0]]
 *     Explanation: The explanation of this example is shown below:
 *     Notice that 0 represents False and 1 represents True in the photo representing the Quad-Tree.
 *
 * Example 2:
 *     Input: grid = [[1,1,1,1,0,0,0,0],[1,1,1,1,0,0,0,0],[1,1,1,1,1,1,1,1],[1,1,1,1,1,1,1,1],[1,1,1,1,0,0,0,0],[1,1,1,1,0,0,0,0],[1,1,1,1,0,0,0,0],[1,1,1,1,0,0,0,0]]
 *     Output: [[0,1],[1,1],[0,1],[1,1],[1,0],null,null,null,null,[1,0],[1,0],[1,1],[1,1]]
 *     Explanation: All values in the grid are not the same. We divide the grid into four sub-grids.
 *     The topLeft, bottomLeft and bottomRight each has the same value.
 *     The topRight have different values so we divide it into 4 sub-grids where each has the same value.
 *     Explanation is shown in the photo below:
 *
 * Constraints:
 *     - n == grid.length == grid[i].length
 *     - n == 2x where 0 <= x <= 6
 *
 * Template (python3):
 *     '''
 *     # Definition for a QuadTree node.
 *     class Node:
 *         def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
 *             self.val = val
 *             self.isLeaf = isLeaf
 *             self.topLeft = topLeft
 *             self.topRight = topRight
 *             self.bottomLeft = bottomLeft
 *             self.bottomRight = bottomRight
 *     '''
 *
 *     class Solution:
 *         def construct(self, grid: List[List[int]]) -> 'Node':
 */

use wasm_libs::*;

fn all_same(grid: &[Vec<i32>], row: usize, col: usize, size: usize) -> bool {
    let v = grid[row][col];
    for i in row..row + size {
        for j in col..col + size {
            if grid[i][j] != v { return false; }
        }
    }
    true
}

enum QNode {
    Leaf(i32),
    Internal(i32, usize),
}

fn main() {
    let header = read_ints();
    let cols = header[0] as usize;
    let mut grid: Vec<Vec<i32>> = Vec::with_capacity(cols);
    for _ in 0..cols {
        grid.push(read_ints());
    }

    let mut flat: Vec<i32> = Vec::new();
    fn build_flat(grid: &[Vec<i32>], row: usize, col: usize, size: usize, flat: &mut Vec<i32>) -> usize {
        if all_same(grid, row, col, size) {
            flat.push(1);
            flat.push(grid[row][col]);
            flat.len() - 2
        } else {
            let half = size / 2;
            let idx = flat.len();
            flat.push(0);
            flat.push(1);
            build_flat(grid, row, col, half, flat);
            build_flat(grid, row, col + half, half, flat);
            build_flat(grid, row + half, col, half, flat);
            build_flat(grid, row + half, col + half, half, flat);
            idx
        }
    }
    build_flat(&grid, 0, 0, cols, &mut flat);

    use std::io::Write;
    let mut out = std::io::stdout().lock();

    fn compute_child_sizes(flat: &[i32], pos: usize) -> usize {
        if flat[pos] == 1 { return 2; }
        let mut total = 2;
        let mut p = pos + 2;
        for _ in 0..4 {
            let cs = compute_child_sizes(flat, p);
            p += cs;
            total += cs;
        }
        total
    }

    fn bfs_print(flat: &[i32], root_pos: usize, out: &mut std::io::StdoutLock) {
        let mut queue = std::collections::VecDeque::new();
        queue.push_back(root_pos);
        while let Some(pos) = queue.pop_front() {
            if flat[pos] == 1 {
                writeln!(out, "{} {}", flat[pos], flat[pos + 1]).unwrap();
            } else {
                writeln!(out, "{} {}", flat[pos], flat[pos + 1]).unwrap();
                let mut child_pos = pos + 2;
                for _ in 0..4 {
                    queue.push_back(child_pos);
                    child_pos += compute_child_sizes(flat, child_pos);
                }
            }
        }
    }

    bfs_print(&flat, 0, &mut out);
    std::process::exit(0);
}
